torch/distributed/_checkpoint/state_dict_saver.py

```diff
@@ -0,0 +1,213 @@
+from typing import Any, Callable, Dict, List, Tuple
+import torch
+from torch import Tensor
+from .metadata import Metadata, ReadWriteRequest, ExtendedTensorMetadata, StorageMetadata
+from .storage_writer import StorageWriter
+from torch.futures import Future
+from torch.distributed.shard.sharded_tensor import (
+    ShardedTensor,
+    ShardedTensorMetadata,
+    TensorProperties,
+    ShardMetadata,
+)
+import torch.distributed as dist
+
+# -------------- private functions --------------
+def _populate_inplace_with_a_tensor(
+    fqn: str,
+    tensor: Tensor,
+    metadata: Metadata,
+    size_for_storage_handles: Dict[str, int],
+    write_requests: List[ReadWriteRequest],
+):
+    # --- Step 1, populate write request ---
+    # The reshaping should not be required here
+    target_tensor = torch.reshape(tensor, (-1,))
+    storage_size = target_tensor.nelement() * target_tensor.element_size()
+
+    wr = ReadWriteRequest(
+        target_tensor=target_tensor,
+        storage_key=fqn,
+        offset=0,
+        length=storage_size,
+    )
+
+    write_requests.append(wr)
+
+    # --- Step 2, populate the size_for_storage_handles ---
+    #
+    size_for_storage_handles[fqn] = storage_size
+
+    # --- Step 3, populate the metadata ---
+    #
+    # Since torch.Tensor does not have a standard set of metadata we can operate on
+    # We wrap troch.Tensor's metadata with ShardMetadata
+    # This is frankly a bad idea, I will need to change this
+    tensor_size = list(tensor.detach().size())
+    shard_metadata = ShardMetadata(
+        shard_offsets=[0] * len(tensor_size),
+        shard_sizes=tensor_size,
+        # Not sure how to deal with placement for regular tensor yet.
+        # Since they are only keep the copy on rank0, let's hard code it for now.
+        placement=f"rank:0/{str(tensor.device)}",
+    )
+    stm = ShardedTensorMetadata(
+        shards_metadata=[shard_metadata],
+        size=tensor.size(),
+        tensor_properties=TensorProperties(
+            dtype=tensor.dtype,
+            layout=tensor.layout,
+            requires_grad=tensor.requires_grad,
+            memory_format=torch.contiguous_format,
+            pin_memory=tensor.is_pinned(),
+        )
+    )
+
+    etmd = ExtendedTensorMetadata(
+        tensor_metadata=stm,
```

Comment:
For the torch.Tensor case, can't tensor_metadata be `Union[ShardedTensorMetadata, TensorProperties]`? Or probably we need to create another class like `TensorMetadata` which includes size as well since TensorProperties doesn't have size.

Reusing ShardedTensorMetadata here doesn't seem appropriate.